src/datasources/db/v2/database-migrator.service.ts

```diff
@@ -0,0 +1,160 @@
+  type ILoggingService,
+} from '@/logging/logging.interface';
+import { Inject, Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import type { DataSource } from 'typeorm';
+
+interface LockSchema {
+  id: number;
+  name: number;
+}
+
+enum MigrationStatus {
+  RUNNING = 1,
+}
+
+@Injectable()
+export class DatabaseMigrator {
+  public constructor(
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+
+    @Inject(PostgresDatabaseService)
+    private readonly databaseService: PostgresDatabaseService,
+
+    @Inject(ConfigService)
+    private readonly configService: ConfigService,
+  ) {
+    //
+  }
+
+  /**
+   * Performs the migration operation.
+   *    Since pg_advisory_lock locks are session-based, they seem to be quite unreliable for our use case.
+   *    If the session is terminated, the lock is released, which could potentially cause issues in the database.
+   *    For that reason we are implementing a lock mechanism
+   *
+   * @returns {Promise<void>} A promise that resolves when the migration is complete.
+   *
+   * @throws {Error} Throws an error if unsuccessful after a specified number of retries."
+   */
+  public async migrate(): Promise<void> {
+    this.loggingService.info('Migrations: Running...');
+
+    const connection =
+      await this.databaseService.initializeDatabaseConnection();
+    await this.createLockTableIfNotExists(connection);
+
+    let numberOfIterations = 0;
+    const numberOfRetries = await this.configService.getOrThrow(
+      'db.migrator.numberOfRetries',
+    );
+    while (numberOfRetries >= numberOfIterations) {
+      ++numberOfIterations;
+      if (!(await this.lockExists(connection))) {
+        await this.insertLock(connection);
+        await this.runMigration(connection);
+        await this.truncateLocks(connection);
+
+        this.loggingService.info('Migrations: Finished.');
+        break;
+      } else {
+        if (numberOfIterations === numberOfRetries) {
+          throw new Error(
+            'Migrations: Migrations are still running in another instance!',
+          );
+        }
+        this.loggingService.info('Migrations: Running in another instance...');
+        const retryAfter = this.configService.getOrThrow<number>(
+          'db.migrator.retryAfter',
+        );
+
+        await new Promise((resolve) => setTimeout(resolve, retryAfter));
+      }
+    }
+  }
+
+  /**
+   * Creates the locks table in the database if it does not already exist.
+   *
+   * @param {DataSource} connection The database connection to create the locks table on.
+   *
+   * @returns {Promise<void>} A promise that resolves when the locks table has been created or confirmed to exist.
+   */
+  private async createLockTableIfNotExists(
+    connection: DataSource,
+  ): Promise<void> {
+    await connection.query(
+      'CREATE TABLE IF NOT EXISTS "_lock" ("id" SERIAL NOT NULL, "status" int NOT NULL, PRIMARY KEY ("id"))',
```

Comment:
I'd suggest moving the table name to a constant, and using it in all methods here. What do you think?

Reply:
Done
https://github.com/safe-global/safe-client-gateway/pull/1988/commits/2ea0e5d8a752fd31cff745ff3490d8fe8134ddc7